src/wasm/adapter.rs

```diff
@@ -0,0 +1,153 @@
+use super::peripheral::Peripheral;
+use crate::api::{BDAddr, Central, CentralEvent, Peripheral as _, PeripheralId};
+use crate::{Error, Result};
+
+use crate::common::adapter_manager::AdapterManager;
+use async_trait::async_trait;
+use std::pin::Pin;
+
+use futures::channel::oneshot;
+use futures::stream::Stream;
+use wasm_bindgen::prelude::*;
+
+use js_sys::Array;
+use wasm_bindgen_futures::{spawn_local, JsFuture};
+use web_sys::{BluetoothDevice, RequestDeviceOptions};
+
+/// Implementation of [api::Central](crate::api::Central).
+#[derive(Clone, Debug)]
+pub struct Adapter {
+    manager: AdapterManager<Peripheral>,
+}
+
+fn bluetooth() -> Option<web_sys::Bluetooth> {
+    web_sys::window().unwrap().navigator().bluetooth()
+}
+
+#[macro_export]
+macro_rules! spawn_local_and_wait {
```

Comment:
Just for reference; I currently depend on being able to convert a BDAddr into a u64 and back for more practical marshalling of addresses across an ffi boundary, between C# and Rust for a Unity plugin.

A more open-ended vec/string representation would make things a little awkward here since it potentially introduces more fiddly memory mangement sending addresses to C#. So long as the address is hashable I could hide this fairly easily by maintaining my own integer addresses that map to the real address but maybe it could even be useful if btleplug would keep backend address representation an implementation detail and always just return a u64 address - and then just leave it up to the backend to map that to something. For winrt, corebluetooth, bluez the native address can be packed in a u64, or wasm a hash table could be used in the backend to map the string ids to unique integers that get handed out to applications?

Arguably that's kind of what we already have I suppose, and I guess wasm could have a backend index of string addresses and look up a BDaddr from that to hand out for the public API. A unique BDAddr can be generated from a monotonically incremented u64 (BDAddr already has conversion to/from u64)

Reply:
1) The macOS implementation already requires the usage of 128bit UUIDs, so the existing implementation is not 100% correct.
2) What's the benefit of maintaining the map in btleplug than in the ffi wrapper? What's the problem with wrapping the new `PeripheralId` in ffi?

Reply:
Oof yeah actually the current uuid mapping for the corebluetooth backend isn't great :)

```rust
pub(crate) fn uuid_to_bdaddr(uuid: &str) -> BDAddr {
    let b: [u8; 6] = uuid.as_bytes()[0..6].try_into().unwrap();
    BDAddr::try_from(b).unwrap()
}
```
That's just taking 48 bits from the _string_ representation of the uuid :)

There's no particularly special benefit to having a mapping like I was suggesting inside btleplug; I was just thinking that if the wasm backend is the only backend that needs anything more than a 64bit ID it could be convenient to maintain that ability to reliably convert addresses to u64. Seeing that the corebluetooth backend can't really convert device IDs to a u64 I'm not so sure about this constraint now.

Looking at the Android API (which I'm hoping to use later) I see they simply report the device address as a string. They document that it's the mac address and imply you can rely on it being in a standard format but it also doesn't look like a really string guarantee and it's maybe not good to assume that the string can be decoded:

```
public String getAddress ()

Returns the hardware address of this BluetoothDevice.

For example, "00:11:22:AA:BB:CC"
```
(https://developer.android.com/reference/android/bluetooth/BluetoothDevice#getAddress())

Tbh for my use case I could also quite easily just not pass the address to Unity/C# since it's not really something that needs to be shown to the user. I create a handle for the peripheral itself which serves as a unique ID for interacting with the api over ffi. If needs be for showing the IDs to the user then passing a string across isn't that big of a deal.